### PR TITLE
Setup github actions to build web ui

### PR DESCRIPTION
**Pull request by mahmednabil109**
- add github actions to build the react app and embed it using esc

Signed-off-by: Mohamed Abokammer <mahmednabil109@gmail.com>

### CODE REVIEW

File: .github/workflows/ui-build.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+          cd "${GITHUB_WORKSPACE}"/src/github.com/${{ github.repository }}/cmds/admin_server/ui/dist
+          go get -u github.com/mjibson/esc
+          go install github.com/mjibson/esc
+          export PATH=$PATH:$(go env GOPATH)/bin
+          # setting modtime to a dummy fixed value to avoid unnecessary commits
+          esc -o ../../server/static.go -pkg=server -modtime=0 -prefix=/ ./
+      - uses: EndBug/add-and-commit@v9
+        with:
+          message: 'embed web/ui into go file'
+          add: '*'
```
Comment (mimir-d) [inline .github/workflows/ui-build.yaml:34]: not all files
```suggestion
          add: -f <path to static.go>
```

File: .github/workflows/ui-build.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+          set -exu
+          cd "${GITHUB_WORKSPACE}"/src/github.com/${{ github.repository }}/cmds/admin_server/ui/dist
+          go get -u github.com/mjibson/esc
+          go install github.com/mjibson/esc
+          export PATH=$PATH:$(go env GOPATH)/bin
+          # setting modtime to a dummy fixed value to avoid unnecessary commits
+          esc -o ../../server/static.go -pkg=server -modtime=0 -prefix=/ ./
+      - uses: EndBug/add-and-commit@v9
+        with:
+          message: 'embed web/ui into go file'
```
Comment (mimir-d) [inline .github/workflows/ui-build.yaml:33]: would also add
`default_author: github_actions`

File: cmds/admin_server/server/server.go [hunk 3]
```diff
@@ -132,13 +131,13 @@ func initRouter(ctx xcontext.Context, rh RouteHandler) *gin.Engine {
...
 
-	// serve the frontend app
-	r.Use(static.Serve("/", static.LocalFile("./frontend/dist", true)))
-
 	r.GET("/status", rh.status)
 	r.POST("/log", rh.addLog)
 	r.GET("/log", rh.getLogs)
 
+	// serve the frontend app
+	r.StaticFS("/app", FS(false))
```
Comment (mimir-d) [inline cmds/admin_server/server/server.go:139]: this needs to be in the root of the server; the rest api should be moved to "/api/v1"; do this in another PR because it's not part of "setup gh actions to build webui"

**Review by mimir-d — CHANGES_REQUESTED** [review]

File: .github/workflows/ui-build.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+name: Web UI Build
+on: [push, pull_request]
+jobs:
+  Build:
+    runs-on: ubuntu-20.04
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          path: src/github.com/${{ github.repository }}
```
[the remark this answers is not in the document]
Comment (mahmednabil109) [reply .github/workflows/ui-build.yaml:9]: not a specific reason, but to specify a working directory

[the remark this answers is not in the document]
Comment (mimir-d) [reply .github/workflows/ui-build.yaml:22]: ok, remove this and then you dont need all of the long paths below, because you'll be in the root of the project

`cd "${GITHUB_WORKSPACE}"/src/github.com/${{ github.repository }}/cmds/admin_server/ui`
turns into
`cd cmds/admin_server/ui`

File: .github/workflows/ui-build.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
+name: Web UI Build
+on: [push, pull_request]
```
[the remark this answers is not in the document]
Comment (mimir-d) [reply .github/workflows/ui-build.yaml:2]: ok, here's the easier way to do this; your filter still runs the gh action for all files and then determines not to continue
```
on:
  push:
    paths:
      - 'path/to/ui/**'
  pull_request:
    paths:
      - 'path/to/ui/**'
```

File: cmds/admin_server/server/server.go (hunk 3 above)
Comment (mahmednabil109) [reply cmds/admin_server/server/server.go:139]: I agree to move the rest api to "/api/v1", but serving static files in the root will cause problems as the static method registers a `/*filepath` which matches all the paths.

**Review by mimir-d — APPROVED** [review]
